packages/next/src/client/components/navigation.ts

```diff
@@ -167,11 +169,25 @@ function getSelectedParams(
 export function useParams(): Params {
   clientHookInServerComponentError('useParams')
   const globalLayoutRouterContext = useContext(GlobalLayoutRouterContext)
-  if (!globalLayoutRouterContext) {
-    // This only happens in `pages`. Type is overwritten in navigation.d.ts
-    return null!
+  const pagesRouter = useContext(RouterContext)
+
+  // When it's under app router
+  if (globalLayoutRouterContext) {
+    return getSelectedParams(globalLayoutRouterContext.tree)
+  }
+
+  // When it's under pages router,
+  if (pagesRouter) {
+    const routeRegex = getRouteRegex(pagesRouter.pathname)
+    const pathParams: Params = {}
+    const keys = Object.keys(routeRegex.groups)
+    for (const key of keys) {
+      pathParams[key] = pagesRouter.query[key]!
+    }
+
+    return pathParams
   }
-  return getSelectedParams(globalLayoutRouterContext.tree)
+  return {}
```

Comment:
`return null`?

Reply:
I don't feel we should still return `null` as app router will always return `{}`, I've dropped the null type of return value

Reply:
`null` would align it better with the `useSearchParams()` and will also make clear when the router is "not ready" per page router `isReady` API.